### PR TITLE
[SPARK-47564][SQL] Always throw FAILED_READ_FILE error when fail to read files

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/44953 to refine the newly added `FAILED_READ_FILE` error. It's better to always throw `FAILED_READ_FILE` error if anything goes wrong during file reading. This is more predictable and easier for users to do error handling. This PR adds sub error classes to `FAILED_READ_FILE` so that users can know what went wrong quicker.

### Why are the changes needed?

better error reporting

### Does this PR introduce _any_ user-facing change?

no, `FAILED_READ_FILE` is not released yet.

### How was this patch tested?

existing tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala [hunk 1]
```diff
@@ -436,7 +436,7 @@ abstract class AvroLogicalTypeSuite extends QueryTest with SharedSparkSession {
       val ex = intercept[SparkException] {
         spark.read.format("avro").load(s"$dir.avro").collect()
       }
-      assert(ex.getErrorClass == "FAILED_READ_FILE")
+      assert(ex.getErrorClass.startsWith("FAILED_READ_FILE"))
```
Comment (MaxGekk) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala:439]: Why don't compare to `FAILED_READ_FILE.NO_HINT`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 2]
```diff
@@ -805,33 +799,34 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase with ExecutionE
       cause = null)
   }
 
-  def fileNotFoundError(e: FileNotFoundException): SparkFileNotFoundException = {
-    new SparkFileNotFoundException(
-      errorClass = "_LEGACY_ERROR_TEMP_2062",
-      messageParameters = Map("message" -> e.getMessage))
+  def fileNotExistError(path: String, e: Exception): Throwable = {
+    new SparkException(
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:803]: Why did you change `SparkFileNotFoundException` to `SparkException`. This can break user's apps, I think. If they catch the exact `SparkFileNotFoundException`, they will miss your `SparkException`. Especially, after you changed the error class too. How should user know about your changes?

File: common/utils/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -1257,6 +1249,31 @@
...
+        ]
+      },
+      "NO_HINT" : {
+        "message" : [
+          ""
+        ]
+      },
+      "PARQUET_COLUMN_DATA_TYPE_MISMATCH" : {
+        "message" : [
+          "Data type mismatches when reading Parquet column <column>: Expected: <expectedType>, Found: <actualType>."
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:1273]: BTW, since we are talking about the parquet format, shouldn't we mention that one type is a logical type, but another is a physical type? Otherwise you compare apples to oranges.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala:439]: We may keep adding more sub error classes to `FAILED_READ_FILE` and I want the tests to be stable.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:803]: Note: at the end, we threw `SparkException` before: https://github.com/apache/spark/pull/44953/files#diff-158a89983d40e2ebba7048dd10bac07b6c4fa03d70ee608923b093b314219f19L254 . This change is to make the behavior closer to before.

File: common/utils/src/main/resources/error/error-classes.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json:1273]: I find the logical vs physical type a bit confusing. How about `Expected Spark type: ..., actual Parquet type ...`.

[the remark this answers is not in the document]
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:1273]: SGTM

**Review by MaxGekk — APPROVED** [review]
Waiting for CI. The test failure is related:
```
[info] SparkThrowableSuite:
[info] - No duplicate error classes (23 milliseconds)
[info] - Error classes are correctly formatted (13 milliseconds)
[info] - SQLSTATE is mandatory (4 milliseconds)
[info] - Error category and error state / SQLSTATE invariants (42 milliseconds)
[info] - Message invariants (10 milliseconds)
[info] - Message format invariants (10 milliseconds)
[info] - Error classes match with document *** FAILED *** (54 milliseconds)
[info]   "...et column `<column>`[. Expected Spark type `<expectedType>`, actual Parquet type] `<actualType>`." did not equal "...et column `<column>`[: Expected: `<expectedType>`, Found:] `<actualType>`." The error class document is not up to date. Please regenerate it by running `SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "core/testOnly *SparkThrowableSuite -- -t \"Error classes match with document\""` (SparkThrowableSuite.scala:333)
```

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @cloud-fan and @yaooqinn @HyukjinKwon for review.
